source/zipkin/zipkin_core_types.cc

```diff
@@ -0,0 +1,236 @@
+  writer.String(ipv4_.c_str());
+  writer.Key(ZipkinJsonFieldNames::ENDPOINT_PORT.c_str());
+  writer.Uint(port_);
+  writer.Key(ZipkinJsonFieldNames::ENDPOINT_SERVICE_NAME.c_str());
+  writer.String(service_name_.c_str());
+  if (isset_ipv6_) {
+    writer.Key(ZipkinJsonFieldNames::ENDPOINT_IPV6.c_str());
+    writer.String(ipv6_.c_str());
+  }
+  writer.EndObject();
+  json_string_ = s.GetString();
+
+  return json_string_;
+}
+
+Annotation::Annotation(const Annotation& ann) {
+  timestamp_ = ann.timestamp();
+  value_ = ann.value();
+  endpoint_ = ann.endpoint();
+  isset_endpoint_ = ann.isSetEndpoint();
+}
+
+Annotation& Annotation::operator=(const Annotation& ann) {
+  timestamp_ = ann.timestamp();
+  value_ = ann.value();
+  endpoint_ = ann.endpoint();
+  isset_endpoint_ = ann.isSetEndpoint();
+
+  return *this;
+}
+
+const std::string& Annotation::toJson() {
+  rapidjson::StringBuffer s;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(s);
+  writer.StartObject();
+  writer.Key(ZipkinJsonFieldNames::ANNOTATION_TIMESTAMP.c_str());
+  writer.Uint64(timestamp_);
+  writer.Key(ZipkinJsonFieldNames::ANNOTATION_VALUE.c_str());
+  writer.String(value_.c_str());
+  writer.EndObject();
+
+  json_string_ = s.GetString();
+
+  if (isset_endpoint_) {
+    Util::mergeJsons(json_string_, endpoint_.toJson(),
+                     ZipkinJsonFieldNames::ANNOTATION_ENDPOINT.c_str());
+  }
+
+  return json_string_;
+}
+
+BinaryAnnotation::BinaryAnnotation(const BinaryAnnotation& ann) {
+  key_ = ann.key();
+  value_ = ann.value();
+  annotation_type_ = ann.annotationType();
+  endpoint_ = ann.endpoint();
+  isset_endpoint_ = ann.isSetEndpoint();
+}
+
+BinaryAnnotation& BinaryAnnotation::operator=(const BinaryAnnotation& ann) {
+  key_ = ann.key();
+  value_ = ann.value();
+  annotation_type_ = ann.annotationType();
+  endpoint_ = ann.endpoint();
+  isset_endpoint_ = ann.isSetEndpoint();
+
+  return *this;
+}
+
+const std::string& BinaryAnnotation::toJson() {
+  rapidjson::StringBuffer s;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(s);
+  writer.StartObject();
+  writer.Key(ZipkinJsonFieldNames::BINARY_ANNOTATION_KEY.c_str());
+  writer.String(key_.c_str());
+  writer.Key(ZipkinJsonFieldNames::BINARY_ANNOTATION_VALUE.c_str());
+  writer.String(value_.c_str());
+  writer.EndObject();
+
+  json_string_ = s.GetString();
+
+  if (isset_endpoint_) {
+    Util::mergeJsons(json_string_, endpoint_.toJson(),
+                     ZipkinJsonFieldNames::BINARY_ANNOTATION_ENDPOINT.c_str());
+  }
+
+  return json_string_;
+}
+
+Span::Span(const Span& span) {
+  trace_id_ = span.traceId();
+  name_ = span.name();
+  id_ = span.id();
+  parent_id_ = span.parentId();
+  annotations_ = span.annotations();
+  binary_annotations_ = span.binaryAnnotations();
+  timestamp_ = span.timestamp();
+  duration_ = span.duration();
+  trace_id_high_ = span.traceIdHigh();
+  isset_ = span.isSet();
+  start_time_ = span.startTime();
+  tracer_ = span.tracer();
```

Comment:
Do we need this copy constructor if it's just a one to one assignment of ints and pointers to const objects ?